lgc/patch/PatchReadfirstlaneOpt.cpp

```diff
@@ -0,0 +1,204 @@
+ * @brief LLPC source file: contains implementation of class lgc::PatchReadfirstlaneOpt.
+ ***********************************************************************************************************************
+ */
+#include "PatchReadfirstlaneOpt.h"
+#include "lgc/Builder.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PipelineState.h"
+#include "llvm/Analysis/LegacyDivergenceAnalysis.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/Support/Debug.h"
+#include <stack>
+
+#define DEBUG_TYPE "lgc-patch-readfirstlane-opt"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Define static members (no initializer needed as LLVM only cares about the address of ID, never its value).
+char PatchReadfirstlaneOpt::ID;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for readfirstlane optimizations.
+FunctionPass *createPatchReadfirstlaneOpt() {
+  return new PatchReadfirstlaneOpt();
+}
+
+PatchReadfirstlaneOpt::PatchReadfirstlaneOpt() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in,out] function : Function that we will optimize readfirstlane insertion place.
+bool PatchReadfirstlaneOpt::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Readfirstlane-Opt\n");
+
+  m_divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+  visit(function);
+
+  const bool changed = pushUpReadfirstlane();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadfirstlaneOpt::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchReadfirstlaneOpt::visitCallInst(CallInst &callInst) {
+  // Collect the intial uniform operands as the start point for pushing the corresponding readfirstlanes higher
+  if (auto intrinsicCall = dyn_cast<IntrinsicInst>(&callInst)) {
+    if (intrinsicCall->getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      if (auto extract = dyn_cast<ExtractElementInst>(callInst.getArgOperand(0)))
+        m_intialUniforms.insert(extract->getOperand(0));
```

Comment:
Why limit it to readfirstlane(extractelement)?

Reply:
There should be no limit on readfirstlane. But there are some tests failed in DescriptorIndexing.exe when I tried to remove the limit. I'm working on the fix.